lib/Runtime/Library/AtomicsObject.cpp

```diff
@@ -0,0 +1,270 @@
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "RuntimeLibraryPch.h"
+
+#define ATOMICS_FUNCTION_BEGIN_WRAPPER(length, methodName) \
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault); \
+        ARGUMENTS(args, callInfo); \
+        ScriptContext* scriptContext = function->GetScriptContext(); \
+        Assert(!(callInfo.Flags & CallFlags_New)); \
+        if (args.Info.Count <= length) \
+        { \
+            JavascriptError::ThrowRangeError(scriptContext, JSERR_Function_LessArguments, _u(methodName)); \
+        } \
+
+
+namespace Js
+{
+    Var AtomicsObject::ValidateSharedIntegerTypedArray(Var typedArray, ScriptContext *scriptContext, bool onlyInt32)
+    {
+        if (!TypedArrayBase::Is(typedArray))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedTypedArrayObject);
+        }
+
+        if (onlyInt32)
+        {
+            if (!Int32Array::Is(typedArray))
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_InvalidOperationOnTypedArray);
+            }
+        }
+        else
+        {
+            if (!(Int8Array::Is(typedArray) || Uint8Array::Is(typedArray) || Int16Array::Is(typedArray) ||
+                Uint16Array::Is(typedArray) || Int32Array::Is(typedArray) || Uint32Array::Is(typedArray)))
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_InvalidOperationOnTypedArray);
+            }
+        }
+
+        TypedArrayBase *typedArrayBase = TypedArrayBase::FromVar(typedArray);
+        ArrayBufferBase* arrayBuffer = typedArrayBase->GetArrayBuffer();
+        if (arrayBuffer == nullptr || !ArrayBufferBase::Is(arrayBuffer) || !arrayBuffer->IsSharedArrayBuffer())
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedSharedArrayBufferObject);
+        }
+
+        return arrayBuffer;
+    }
+
+    uint32 AtomicsObject::ValidateAtomicAccess(Var typedArray, Var requestIndex, ScriptContext *scriptContext)
+    {
+        int32 accessIndex = -1;
+        if (TaggedInt::Is(requestIndex))
+        {
+            accessIndex = TaggedInt::ToInt32(requestIndex);
+        }
+        else
+        {
+            accessIndex = JavascriptConversion::ToInt32_Full(requestIndex, scriptContext);
+            double dblValue = JavascriptConversion::ToNumber(requestIndex, scriptContext);
+            if (dblValue != accessIndex)
+            {
+                JavascriptError::ThrowRangeError(scriptContext, JSERR_InvalidTypedArrayIndex);
+            }
+        }
+
+        Assert(TypedArrayBase::Is(typedArray));
+        
+        if (accessIndex < 0 || accessIndex >= (int32)TypedArrayBase::FromVar(typedArray)->GetLength())
```

Comment:
I have test cases for that. Will you check that and verify if that is what you meant?

---
In reply to: [77403224](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77403224) [](ancestors = 77403224)